### DISCUSSION TITLE
Please add Julia!!

### DISCUSSION THREAD

**Discussion by FraCpl** [casadi/casadi#3557] [Ideas]
Hi there,

I've been using Casadi on Matlab for many years now to optimize spacecraft trajectories (I've been working in the space industry and agencies for a while, and still do): IMHO it is the best tool for AD and NLP interfacing.
I am now transitioning from Matlab to Julia... and, even if an unofficial/unregistered casadi.jl package exists here on GitHub, it will be awesome if we could get an official implementation. I found out that, for my scope of work, Casadi still outperforms any AD engine I've tried on Julia.

So... please consider developing an official Julia package!

Cheers

**Comment by jaeandersson** [comment]
It could definitely be done, but it's a lot of work to develop and maintain a new front end. The easiest way is probably to add a SWIG backend to Julia, which could be useful in other projects as well. Anyway, there would need to be a pull from some project with associated funding.

**Reply by jaeandersson** [reply]
Another interesting option could be to implement symbolic export from Julia to CasADi.

**Reply by nashmit** [reply]
If you can go into more details ( perhaps there is something specific you would like to be taken into consideration ) there might be a chance that someone else would want to get involved. I for one would be interested in giving it a try.

**Reply by chelseas** [reply]
Julia has the type `Expr` which is very flexible for building symbolic expressions. They can be traversed: `myexpr.args = :x, :y`

**Reply by jaeandersson** [reply]
Exporting from Julia to CasADi mainly requires knowledge of traversing Julia expressions. You should for example be able to construct a CasADi expression from inside Julia via its Python interoperability and then serialize the CasADi expressions to use elsewhere.

**Reply by pekpuglia** [reply]
The main issue with the Python interfaces (such as in the unofficial https://github.com/ichatzinikolaidis/CasADi.jl) is the speed/memory usage. I have Casadi directly from MATLAB and with this stack of Julia/PyCall/CasADi and I can say that the experience of using CasADi through Python is much worse. 

Not to mention some weirdness that happens with the SX type, which is not properly interpreted neither as scalar neither as vector in Julia through the PyCall interface. 

Wouldn't a tool like [CxxWrap.jl](https://github.com/JuliaInterop/CxxWrap.jl) make this relatively straightforward?

**Reply by jaeandersson** [reply]
> Wouldn't a tool like [CxxWrap.jl](https://github.com/JuliaInterop/CxxWrap.jl) make this relatively straightforward?

Anyone who wants to give it a shot is more than welcome to do so! Given the current improving support for ONNX import in CasADi, I would also consider Julia -> ONNX -> CasADi

**Reply by jgillis** [reply]
Yeah I think we ONNX would be ideal to map back and forth from Jump and casadi, and will be in casadi main soon.

For access to to the casadi API, swig generating
cxxwrap.jl instructions could be an option.

The missing thing here is not ideas/tech but development time (ideally, sponsored).

**Comment by FraCpl** [comment]
Just for information, I ended up using PythonCall.jl and building a very simple interface that does the job (for the very specific application I'm dealing with).

Thank you all for the suggestions

```julia
# Basic math on Py - CasADi interface
fabs(x::Py) = casadi[].fabs(x)
Base.sin(x::Py) = casadi[].sin(x)
Base.cos(x::Py) = casadi[].cos(x)
Base.tan(x::Py) = casadi[].tan(x)
Base.max(x::Py, y::Py) = casadi[].fmax(x, y)
Base.min(x::Py, y::Py) = casadi[].fmin(x, y)
Base.exp(x::Py) = casadi[].exp(x)
Base.sqrt(x::Py) = casadi[].sqrt(x)
Base.sincos(x::Py) = (sin(x), cos(x))
Base.zero(x::Py) = 0 * x
Base.transpose(a::Matrix{Py}) = [a[j, i] for i in 1:size(a, 1), j in 1:size(a, 2)]
Base.transpose(x::Vector{Py}) = reshape(x, 1, length(x))
Base.adjoint(x::Vector{Py}) = reshape(x, 1, length(x))

LinearAlgebra.dot(x::Vector{Py}, y::Vector{Py}) = sum(x .* y)
LinearAlgebra.dot(x::AbstractVector, y::Vector{Py}) = sum(x .* y)
LinearAlgebra.dot(x::Vector{Py}, y::AbstractVector) = sum(x .* y)
LinearAlgebra.norm(x::Vector{Py}) = casadi[].sqrt(dot(x, x))

# NLPSOL interface
function nlpsol(
    solver::String,
    eval_f::Function,
    eval_g::Function,
    x0::Vector{T},
    xlb::Vector{T},
    xub::Vector{T},
    glb::Vector{T},
    gub::Vector{T};
    opts=Dict(),    
) where {T}

    x = casadi[].SX.sym("x", length(x0))
    X = [x[k - 1] for k in 1:length(x0)]    
    nlp = pydict(; x=x, f=eval_f(X), g=casadi[].vertcat(eval_g(X)...))
    S = casadi[].nlpsol("S", solver, nlp, pydict(opts))
    sol = S(; x0=x0, lbx=xlb, ubx=xub, lbg=glb, ubg=gub)
    x_opt = pyconvert(Vector{T}, sol["x"].elements())
    f_opt = pyconvert(T, sol["f"].elements()[0])

    return x_opt, f_opt
end
```
